KAFKA-14991: Implementation of KIP-937 which improves message timesta…

mehbey opened the pull request:
KIP-937 Implementation
On a high level this implementation introduces two new configurations `log.message.timestamp.before.max.ms` and `log.message.timestamp.after.max.ms` and  deprecates `log.message.timestamp.difference.max.ms`

The default value for all these three configs is maintained to be Long.MAX_VALUE for backward compatibility but with the newly added configurations we can have a finer control when validating message timestamps that are in the past and the future compared to the broker's timestamp.

*More detailed description of your change,
- A message with past timestamp compared to the broker timestamp will be validated using `log.message.timestamp.before.max.ms`
- A message with future timestamp compared to the broker timestamp will be validated using `log.message.timestamp.after.max.ms`
- If these validations fail we return return Error code 32 (INVALID_TIMESTAMP) - Existing behavior no change in error message 
- We have deprecated `log.message.timestamp.difference.max.ms` and it is planned for removal in 4.0
- To maintain backward compatibility if  the default value of `log.message.timestamp.before.max.ms` is not changed, we are assuming customer's are still using the deprecated config `log.message.timestamp.difference.max.ms` and validation is done using its value. This ensures that existing customers who have customized the value of `log.message.timestamp.difference.max.ms` will continue to see no change in behavior


*Summary of testing strategy (including rationale)
- Unit test coverage
- Functional Test cases
    - Set no custom configuration and verify that default values are taking effect - verified through trace logs
    - Set configs `log.message.timestamp.before.max.ms` and `log.message.timestamp.before.max.ms` in and validated that the Before and After validation works as expected and messages are rejected with the expected error message
    - Dynamically overwritten `message.timestamp.after.max.ms` and `message.timestamp.before.max.ms` on a topic and verified that new configuration is taking effect.
    - Set config `log.message.timestamp.difference.max.ms` and verified that it has precedence over `log.message.timestamp.before.max.ms` config - Backward compatibility test case

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

divijvaidya commented on core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala:
```diff
@@ -124,7 +124,7 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
   @ValueSource(strings = Array("zk", "kraft"))
   def testSendWithInvalidCreateTime(quorum: String): Unit = {
```
let's retain the test with older/deprecated config and two new tests with max and min config here.

mehbey replied:
Ack - will update everywhere

divijvaidya commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -642,7 +642,7 @@ class DynamicBrokerReconfigurationTest extends QuorumTestHarness with SaslSetup
     props.put(KafkaConfig.CompressionTypeProp, "gzip")
     props.put(KafkaConfig.LogPreAllocateProp, true.toString)
     props.put(KafkaConfig.LogMessageTimestampTypeProp, TimestampType.LOG_APPEND_TIME.toString)
-    props.put(KafkaConfig.LogMessageTimestampDifferenceMaxMsProp, "1000")
```
we need to retain all tests with older config so that we can catch regression in older config in 3.x series. Please add new tests for new configs everywhere

mehbey replied:
I have made three types of changes to teh tests
1) Added parametrized tests to validation on all new configs and the deprecated one. See changes in `PlaintextProducerSendTest.scala` and `ProducerRequestTest.scala`
2) Added test cases for backward compatibility for both `before` and `after` cases
3) Removed unnecessary calls to set `TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG` to the default value. This calls don't add any value to the test since the calls are not changing the default value. See example changes in `LogCleanerTest.scala` and `LogManagerTest.scala`

clolov commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1862,7 +1884,23 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+  @deprecated("3.6")
   def logMessageTimestampDifferenceMaxMs: Long = getLong(KafkaConfig.LogMessageTimestampDifferenceMaxMsProp)
+
+  //In the transition period before logMessageTimestampDifferenceMaxMs is removed, to maintain backward compatibility,
```
```suggestion
  // In the transition period before logMessageTimestampDifferenceMaxMs is removed, to maintain backward compatibility,
```

clolov commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -289,8 +300,12 @@ public Optional<String> serverConfigName(String configName) {
-                atLeast(0), MEDIUM, TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_DOC)
+            .define(MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG, LONG, DEFAULT_MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS,
+                atLeast(0), MEDIUM, MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_DOC)
+            .define(TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG, LONG, DEFAULT_MESSAGE_TIMESTAMP_BEFORE_MAX_MS,
+                        atLeast(0), MEDIUM, TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_DOC)
+            .define(TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_CONFIG, LONG, DEFAULT_MESSAGE_TIMESTAMP_AFTER_MAX_MS,
+                        atLeast(0), MEDIUM, TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_DOC)
```
```suggestion
            .define(TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG, LONG, DEFAULT_MESSAGE_TIMESTAMP_BEFORE_MAX_MS,
                atLeast(0), MEDIUM, TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_DOC)
            .define(TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_CONFIG, LONG, DEFAULT_MESSAGE_TIMESTAMP_AFTER_MAX_MS,
                atLeast(0), MEDIUM, TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_DOC)
```

clolov commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -386,13 +406,27 @@ public LogConfig(Map<?, ?> props, Set<String> overriddenConfigs) {
         this.messageTimestampType = TimestampType.forName(getString(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG));
         this.messageTimestampDifferenceMaxMs = getLong(TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG);
+        this.messageTimestampBeforeMaxMs = getMessageTimestampBeforeMaxMs();
+        this.messageTimestampAfterMaxMs = getLong(TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_CONFIG);
```
Maybe this was called out elsewhere - I will go through the KIP + email discussion - but shouldn't this also default to the specified value of `log.message.timestamp.difference.max.ms` (default or otherwise) if `log.message.timestamp.before.max.ms` is not set?

If not do we not risk the following situation?
* A customer has specified a value of `log.message.timestamp.difference.max.ms` to 100ms.
* The customer upgrades their Kafka version.
* The customer does not set `log.message.timestamp.before.max.ms`.
* There is a misbehaving producer which previously sent records with CreateTimestamp > 10000ms in the future.
* `log.message.timestamp.after.max.ms` defaults to the gigantic number since it is not set and messages from that producer start being accepted while they were previously rejected.

mehbey replied:
Good call out @clolov. To make the transition smooth for customers, we will need to check if customer has changed the value of for both `log.message.timestamp.before.max.ms` and `log.message.timestamp.after.max.ms` and fall back to `log.message.timestamp.difference.max.ms` if it is not set.

I will update with the PR, thank you.

clolov reviewed (commented):
Thank you! I made a first pass and left a couple of comments

mehbey committed:
KAFKA-14991: Resolved merge conflict, Added parameterized tests to cover more scenarios. Addressed PR comments

mehbey committed:
KAFKA-14991: Implementation of KIP-937 which improves message timestamp validation. Detail changes 1) added two new configurations log.message.timestamp.before.max.ms and message.timestamp.after.max.ms. 2) Deprecated log.message.timestamp.difference.max.ms

mehbey commented:
Updated PR to address comments

clolov reviewed (approved):
Great, thank you. The changes make sense to me!

As far as I can tell, the test failures are unrelated.

mehbey committed:
KAFKA-14991: Implementation of KIP-937.  Fixed CI error on JDK 8 and Scala 2.12 build/

divijvaidya commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1862,7 +1884,35 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+  @nowarn("cat=deprecation")
+  def logMessageTimestampBeforeMaxMs: Long = {
+    val messageTimestampBeforeMaxMs: Long = getLong(KafkaConfig.LogMessageTimestampBeforeMaxMsProp)
+    if (messageTimestampBeforeMaxMs != Long.MaxValue) {
```
Instead of `Long.MaxValue` could we use `LogConfig.DEFAULT_MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS`? I know that the value is same but the latter is semantically correct since we want to check for default value.

mehbey replied:
yeah that make sense - updated.

divijvaidya commented on core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala:
```diff
@@ -120,17 +121,17 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
     }
   }
 
-  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
```
we are not running this test with both kraft and Zk by removing this. Could we please add it back?

mehbey replied:
`quorum: String` this parameter was not used in the test. There are similar unit tests in this file which defines parameters like this @ValueSource(strings = Array("zk", "kraft")) but don't actually use them in the test. I have refrained from cleaning all the tests not to pollute this PR but that can be a quick follow up PR separately.

If there is some reason which I missed but we still want to the parametrized value, we can add it to the argument provider method like 

```
 Arguments.of("zk", TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() - fiveMinutesInMs)),
      Arguments.of("zk", TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() - fiveMinutesInMs)),
      Arguments.of("zk", TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() + fiveMinutesInMs)),
 Arguments.of("kraft", TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() - fiveMinutesInMs)),
      Arguments.of("kraft", TopicConfig.MESSAGE_TIMESTAMP_BEFORE_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() - fiveMinutesInMs)),
      Arguments.of("kraft", TopicConfig.MESSAGE_TIMESTAMP_AFTER_MAX_MS_CONFIG, Long.box(System.currentTimeMillis() + fiveMinutesInMs)),
```

divijvaidya replied:
yeah, that part of the code is a but confusing. You see, we add an annotation `TestWithParameterizedQuorumName`. This annotation basically, changes the name of the test and adds quorum={0} to it. In addition `@ValueSource(strings = Array("zk", "kraft"))` sets the value of test name to be `testSendWithInvalidCreateTime.quorum=zk` or `testSendWithInvalidCreateTime.quorum=kraft`. After the test name has been modified at runtime, `TestInfoUtils.isKRaft` looks at the test name and determines if the test should be run as Zk test or a kraft test. That is why it may seem that this `quorum` param is unnecessary but actually it isn't.

Now, adding this to the argument provider should work, as long as it is the first argument.

mehbey replied:
Thank you for clarifying that. I have added the `quorum` param and verified that the test name is updated during runtime as you described. Here is what I have captured from the test run out put
```
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=zk PASSED
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=zk PASSED
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=zk PASSED
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=kraft PASSED
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=kraft PASSED
Gradle Test Run :core:test > Gradle Test Executor 3 > PlaintextProducerSendTest > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long) > testSendWithInvalidBeforeAndAfterTimestamp(String, String, long).quorum=kraft PASSED
```

divijvaidya commented on core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala:
```diff
@@ -120,17 +121,17 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
-  def testSendWithInvalidCreateTime(quorum: String): Unit = {
+  @ParameterizedTest
+  @MethodSource(Array("timestampConfigProvider"))
+  def testSendWithInvalidBeforeAndAfterTimestamp(messageTimeStampConfig: String, recordTimestamp: Long): Unit = {
```
please modify the "valid" test as well to add the usage of this new configuration which should test for parameters where value is equal to timestamp etc.

mehbey replied:
added more tests cases for valid scenarios of 1) when topic config threshold is the same us the record timestamp 2) when record timestamp is within range of threshold

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogValidatorTest.scala:
```diff
@@ -1398,6 +1437,73 @@ class LogValidatorTest {
+        CompressionType.GZIP,
+        CompressionType.GZIP,
+        false,
+        RecordBatch.MAGIC_VALUE_V1,
```
is there a reason that we are using V1 here? It's the old one and current one is V2.

mehbey replied:
no reason, just a bad copy paste from other tests :). Updated, thank you.

divijvaidya commented on core/src/test/scala/unit/kafka/log/LogValidatorTest.scala:
```diff
@@ -1398,6 +1437,73 @@ class LogValidatorTest {
+  def testRecordWithFutureTimestampIsRejected(): Unit = {
+    val timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L //24 hrs
+    val timestampAfterMaxConfig = 1 * 60 * 60 * 1000L //1 hr
+    val now = System.currentTimeMillis()
```
In the tests we usually use MockTime. Is it possible to use that here?

mehbey replied:
that will make the test a little bit involved. The reason is we are using a producer from the common test until `TestUtils.createProducer` which spins up a test producer using the System's current timestamp. We can potentially refactor that code so that it takes a mock time. But without that the timestamp validation will use the system's current timestamp hence the tests follow the same pattern as well.

divijvaidya commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java:
```diff
@@ -186,6 +191,12 @@ public Optional<String> serverConfigName(String configName) {
     @SuppressWarnings("deprecation")
     private static final String MESSAGE_FORMAT_VERSION_CONFIG = TopicConfig.MESSAGE_FORMAT_VERSION_CONFIG;
 
+    @SuppressWarnings("deprecation")
```
what "deprecation" warning are we suppressing here? Wouldn't these configuration continue to exist in future?

mehbey replied:
This is for `TopicConfig.MESSAGE_TIMESTAMP_DIFFERENCE_MAX_MS_CONFIG` It will be removed in the 4.0. I have added a comment to make it clear.

divijvaidya commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -761,6 +761,7 @@ class KafkaConfigTest {
   }
 
   @Test
+  @nowarn("cat=deprecation")
```
please add a comment where ever we are adding this on which particular part of code is deprecated. It helps to ensure that this annotation is deleted when we remove the deprecated code.

mehbey replied:
Updated everywhere. Thank you.

divijvaidya reviewed (commented):
Whenever we are making customer facing changes (adding new config in this scenario), we usually mention it in upgrade notes at https://github.com/apache/kafka/blob/ddeb89f4a9f31b5ff63661e18a94c403a8f45f69/docs/upgrade.html#L24 

Please add information about change made in this PR over there.

mehbey committed:
KAFKA-14991: Implementation of KIP-937.  Addressed PR comments.

mehbey committed:
KAFKA-14991: Implementation of KIP-937.  Addressed more PR comments.

mehbey commented:
> Whenever we are making customer facing changes (adding new config in this scenario), we usually mention it in upgrade notes at
> 
> https://github.com/apache/kafka/blob/ddeb89f4a9f31b5ff63661e18a94c403a8f45f69/docs/upgrade.html#L24
> 
> Please add information about change made in this PR over there.

Updated documentation to indicate that we have deprecated ` log.message.timestamp.difference.max.ms ` config and added two new configs.

divijvaidya reviewed (approved):
Thank you Mehari for addressing all the comments! Change looks good to me. I am waiting for CI to be stable before this in.

divijvaidya committed:
Merge branch 'trunk' into KAFKA-14991-Implement-KIP-937

divijvaidya commented on core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala:
```diff
@@ -121,16 +122,18 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
   @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
-  @ValueSource(strings = Array("zk", "kraft"))
-  def testSendWithInvalidCreateTime(quorum: String): Unit = {
+  @MethodSource(Array("quorumAndTimestampConfigProvider"))
```
@mehbey this test is failing in CI with errors and it looks related to this change. Please look into it when you get a chance.


https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/9/testReport/junit/kafka.api/PlaintextProducerSendTest/Build___JDK_17_and_Scala_2_13___initializationError/

mehbey replied:
That was an oversight from my side when changing the parameter provider to include `quorum`. Updated and verified that all tests are passing in this file. Will check back again to verify that CI is happy.

mehbey committed:
KAFKA-14991: Implementation of KIP-937.  Fixed failing test from wrong parameter provider initialization.

divijvaidya commented:
Unrelated test failures:
```
[Build / JDK 20 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_20_and_Scala_2_13___testThreadPoolResize__/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.OffloadAndConsumeFromLeaderTest.initializationError](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.tiered.storage.integration/OffloadAndConsumeFromLeaderTest/Build___JDK_20_and_Scala_2_13___initializationError/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.streams.integration.NamedTopologyIntegrationTest.shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.streams.integration/NamedTopologyIntegrationTest/Build___JDK_20_and_Scala_2_13___shouldRemoveOneNamedTopologyWhileAnotherContinuesProcessing__/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_20_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 17 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_17_and_Scala_2_13___testRackAwareRangeAssignor__/)
[Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.api/TransactionsTest/Build___JDK_17_and_Scala_2_13___testBumpTransactionalEpoch_String__quorum_kraft/)
[Build / JDK 17 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_17_and_Scala_2_13___testThreadPoolResize__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_17_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_17_and_Scala_2_13___testBalancePartitionLeaders___2/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.OffloadAndConsumeFromLeaderTest.initializationError](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.tiered.storage.integration/OffloadAndConsumeFromLeaderTest/Build___JDK_17_and_Scala_2_13___initializationError/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_17_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsWithCustomForwardingAdminIntegrationTest/Build___JDK_11_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 11 and Scala 2.13 / kafka.api.PlaintextAdminIntegrationTest.testCreatePartitions(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.api/PlaintextAdminIntegrationTest/Build___JDK_11_and_Scala_2_13___testCreatePartitions_String__quorum_zk/)
[Build / JDK 11 and Scala 2.13 / kafka.server.DelegationTokenRequestsTest.testDelegationTokenRequests(String).quorum=kraft](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.server/DelegationTokenRequestsTest/Build___JDK_11_and_Scala_2_13___testDelegationTokenRequests_String__quorum_kraft/)
[Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_11_and_Scala_2_13___testThreadPoolResize__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders___2/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.OffloadAndConsumeFromLeaderTest.initializationError](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14135/10/testReport/junit/org.apache.kafka.tiered.storage.integration/OffloadAndConsumeFromLeaderTest/Build___JDK_11_and_Scala_2_13___initializationError/)
```

divijvaidya commented:
Merging to 3.6 as the KIP was targeted for 3.6 in release plan.
